WikiWatershed/mmw-geoprocessing
# Only download tiles once

Instead of downloading the tiles once for every (multi)polygon in the input list, the tiles are now only downloaded once.

This was achieved by reorganizing the code so that the loop over the RDDs of tiles is the outer-most loop.

Connects #7

**To test**
1. Setup and run `spark-jobserver` (see the "old instructions" section of the P/R description of https://github.com/WikiWatershed/model-my-watershed/pull/713  for instructions).
2. Repeat the following steps for this branch, well as dbbbc18f4de (which is the current head of `develop` at the time of writing).
   1. Compile the code by typing `sbt compile assembly` from within the directory.
   2. Submit the jar to SJS:
      - `curl --data "" 'http://localhost:8090/contexts/summary-context'`
      - `curl --data-binary @summary/target/scala-2.10/mmw-geoprocessing-assembly-0.1.1.jar 'http://localhost:8090/jars/summary'`
   3. Run the following command five or six times to warm up your JVM: `curl --data-binary @examples/request.json 'http://localhost:8090/jobs?sync=true&context=summary-context&appName=summary&classPath=org.wikiwatershed.mmw.geoprocessing.SummaryJob'`
   4. Now run the same command a few times, this time taking timings: `time curl --data-binary @examples/request.json 'http://localhost:8090/jobs?sync=true&context=summary-context&appName=summary&classPath=org.wikiwatershed.mmw.geoprocessing.SummaryJob'`.  On real hardware, this typically takes a little under one second for me.
   5. Now attempt a query that asks for the histograms of seven identical polygons: `time curl --data-binary @examples/request-7x.json 'http://localhost:8090/jobs?sync=true&context=summary-context&appName=summary&classPath=org.wikiwatershed.mmw.geoprocessing.SummaryJob'`.  The old version of the code will take more than 10 seconds, so the synchronous request will time out.  On real hardware, the new version of the code will complete in about one second.

**Another way to test**
Overwrite the old geoprocessing jar in the `/opt/geoprocessing` directory of the worker VM with the jar generated by this code, then reload the worker and try the site.
